refactor: move stores to crates

igamigo opened the pull request.

tomyrd commented on crates/indexed-db-store/src/lib.rs:
```diff
@@ -0,0 +1,54 @@
+use wasm_bindgen::prelude::wasm_bindgen;
+use wasm_bindgen_futures::{JsFuture, js_sys};
+
+// TODO: This functionality is not directly related to the webstore. As such, it should be moved
+// into the webclient crate specifically.
```
Is this TODO new? Should this be addressed in this PR?

igamigo replied:
It's new. I was thinking this PR is already a bit complicated so I wanted to leave it for a follow-up (also, it doesn't block anything). The main thing is that the IndexedDB store ends up containing functionality for the keystore, but in reality they should be self-contained things.  Some of this may be addressed in https://github.com/0xMiden/miden-client/pull/1267

tomyrd commented on crates/testing/miden-client-tests/src/tests.rs:
```diff
@@ -544,23 +393,24 @@ async fn sync_state_mmr() {
-    let mmr_proof = partial_mmr.open(4).unwrap().unwrap();
-    let (block_4, _) = rpc_api.get_block_header_by_number(Some(4.into()), false).await.unwrap();
-    partial_mmr.peaks().verify(block_4.commitment(), mmr_proof).unwrap();
+    // TODO: re-enable this
+    // let partial_mmr = client.test_store().build_current_partial_mmr().await.unwrap();
+    // assert!(partial_mmr.forest().num_leaves() >= 6);
+    // assert!(partial_mmr.open(0).unwrap().is_none());
+    // assert!(partial_mmr.open(1).unwrap().is_some());
+    // assert!(partial_mmr.open(2).unwrap().is_none());
+    // assert!(partial_mmr.open(3).unwrap().is_none());
+    // assert!(partial_mmr.open(4).unwrap().is_some());
+    // assert!(partial_mmr.open(5).unwrap().is_none());
+
+    // // Ensure the proofs are valid
+    // let mmr_proof = partial_mmr.open(1).unwrap().unwrap();
+    // let (block_1, _) = rpc_api.get_block_header_by_number(Some(1.into()), false).await.unwrap();
+    // partial_mmr.peaks().verify(block_1.commitment(), mmr_proof).unwrap();
+
+    // let mmr_proof = partial_mmr.open(4).unwrap().unwrap();
+    // let (block_4, _) = rpc_api.get_block_header_by_number(Some(4.into()), false).await.unwrap();
+    // partial_mmr.peaks().verify(block_4.commitment(), mmr_proof).unwrap();
```
Isn't this initialization needed for some of the tests? How are they passing without this enabled?

tomyrd replied:
Oops sorry, I confused this part with `create_prebuilt_mock_chain` for some reason 😅

tomyrd commented on crates/miden-client-sqlite-store/src/sqlite_store/mod.rs:
```diff
@@ -26,33 +40,17 @@ use miden_objects::asset::{Asset, AssetVault};
 use miden_objects::block::{BlockHeader, BlockNumber};
 use miden_objects::crypto::merkle::{InOrderIndex, MerklePath, MerkleStore, MmrPeaks};
 use miden_objects::note::{NoteTag, Nullifier};
-use miden_tx::utils::sync::RwLock;
```
Shouldn't we still use the miden exposed RwLock even if the crate supports std?

igamigo replied:
I changed it to the one exposed by Miden. I don't think it makes a difference in this case and we might consider changing it in the future, but for consistency it should be OK.

igamigo commented on crates/indexed-db-store/src/web_store/transaction/utils.rs:
```diff
@@ -57,6 +57,7 @@ pub async fn insert_proven_transaction_data(
         block_num: executed_transaction.block_header().block_num(),
         submission_height,
         expiration_block_num: executed_transaction.expiration_block_num(),
+        // TODO: make Store::get_current_timestamp() not take &self
```
Ah yes, this was my idea but it would be a problem because we have `Arc<dyn Store>`

igamigo commented on bin/integration-tests/src/tests/config.rs:
```diff
@@ -75,10 +68,14 @@ impl ClientConfig {
             format!("failed to create keystore at path: {}", auth_path.to_string_lossy())
         })?;
 
+        let sqlite_store = SqliteStore::new(store_config)
+            .await
+            .with_context(|| "failed to create sqlite store")?;
+
```
Yeah, good point. Going to roll it back. I think this was from when I originally introduced the extension trait for having a `sqlite_store` which I eventually rolled back.

bobbinth commented on crates/miden-client-sqlite-store/src/lib.rs:
```diff
@@ -0,0 +1,5 @@
+// SQLite-backed Store implementation for miden-client.
+// This crate provides `SqliteStore` and its full implementation.
```
Could we change these into module-level comments (i.e., `//!`) and also expand them a bit to explain the basics about SQLite store?

bobbinth commented on crates/miden-client-sqlite-store/src/lib.rs:
```diff
@@ -0,0 +1,5 @@
+// SQLite-backed Store implementation for miden-client.
+// This crate provides `SqliteStore` and its full implementation.
+
+mod sqlite_store;
+pub use sqlite_store::SqliteStore;
```
Why do we need this level of nesting? could we not get rid of the `sqlite_store` namespace and move everything up one level?

bobbinth commented on crates/miden-client-sqlite-store/src/sqlite_store/mod.rs:
```diff
@@ -40,19 +35,23 @@ use super::{
-use crate::transaction::{TransactionRecord, TransactionStoreUpdate};
+use miden_client::sync::{NoteTagRecord, StateSyncUpdate};
+use miden_client::transaction::{TransactionRecord, TransactionStoreUpdate};
+use miden_client::utils::RwLock;
```
SQLite works only in the `std` context, right? If so, we could probably just use `RwLock` from the Rust standard library.

bobbinth commented on crates/indexed-db-store/Cargo.toml:
nit: should the directory name for this crate be `idxdb-store`?

bobbinth commented on crates/indexed-db-store/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+description            = "IndexedDB-backed Store implementation for miden-client (WASM)"
+edition.workspace      = true
+license.workspace      = true
+name                   = "indexed-db-store"
```
I think the name should probably be `miden-idxdb-store`.

bobbinth commented on crates/indexed-db-store/src/lib.rs:
```diff
@@ -0,0 +1,54 @@
+/// IndexedDB-backed Store implementation for miden-client (WASM)
+mod web_store;
+use serde::{Deserialize, Serialize};
+pub use web_store::WebStore;
```
Similar question as for the SQLite create: is there a reason to have this level of nesting vs. just bringing the modules up by one level?

bobbinth commented on crates/rust-client/README.md:
```diff
@@ -14,19 +14,20 @@ miden-client = { version = "0.11" }
 | ------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------- |
-| `idxdb`      | Includes `WebStore`, an IndexedDB implementation of the `Store` trait. **Disabled by default.**                                                          |
-| `sqlite`     | Includes `SqliteStore`, a SQLite implementation of the `Store` trait. This relies on the standard library. **Disabled by default.**                                                           |
+| `idxdb`      | Enables WebAssembly (browser) support. The Web store implementation now lives in the separate `indexed-db-store` crate. |
```
Why do we still need `idxdb` feature here?

bobbinth reviewed (commented):
Looks good! Thank you! Not a super deep review from me, but I left some comments inline.

Also, a couple additional comments:

1. Let's read some basic README files to the new crates.
2. Let's add the new crates to the crate publish script.

igamigo commented on crates/idxdb-store/src/lib.rs:
```diff
@@ -0,0 +1,265 @@
+
+// Initialize IndexedDB
+#[wasm_bindgen(module = "/src/js/schema.js")]
+extern "C" {
```
No references to the `/src/web_store` path exist

igamigo commented:
I think I addressed all comments. I have a separate PR for dealing with `tonic` features: https://github.com/0xMiden/miden-client/pull/1268. I have not made the docs/readme too extensive but that can be addressed separately. 

> I wonder if the crate dependency path is the right one here. Conceptually, the client depends on the store, not the other way around. Aside from a few of the imports from miden_client that could probably be done away with, the main blocker for this obviously be the Store trait. If this became its own crate rather than a module within miden_client (say, miden-store-traits? A little similar to [RustCrypto/traits](https://github.com/RustCrypto/traits)), then we could remove this unnatural dependency path (it goes in the direction of ports-adapters pattern)

I think this would work well for removing the testing workaround, but I don't think it's strictly necessary either (basically don't see too many other upsides, and we already have a similar situation in `miden-base` with the `miden-testing` which hosts most tests and has not been particularly problematic). I can create an issue to discuss this but I'm thinking on merging this PR for now, and addressing extra changes separately cc @bobbinth

bobbinth commented:
> I think this would work well for removing the testing workaround, but I don't think it's strictly necessary either (basically don't see too many other upsides, and we already have a similar situation in `miden-base` with the `miden-testing` which hosts most tests and has not been particularly problematic). I can create an issue to discuss this but I'm thinking on merging this PR for now, and addressing extra changes separately cc @bobbinth

Yes, agreed. I'd create a separate issue for this and if we decide to take this route, we can do so in a different PR.
